### DISCUSSION TITLE
x-show transition delay or parent transitions to wait for child's

### DISCUSSION THREAD

**Discussion by MuzafferDede** [alpinejs/alpine#668] [4. General]
I have a use case where parent elements needs to wait child elements to finish its transition than parent start its own transition depend on the same variable so we could have a transition chain starting from child to parent. 

I was playing with the idea of having a modifier on transitions as `x-show.delay.1000` to make transition delay manually assigning  value to `el.transitionDelay` but not sure why delay works but transition doesn't. It shows and hides with the delay but just no transitions. 

Thinking if it's possible to assign parent element's transition delay to its total of child element's transition durations. I know it's sounds crazy but it might bring new taste to transitions. 

## Ideas:
- In this example parent will start transition after its child finishes which is 1 second later so it will complete in 2 seconds in total. 
```html
<div x-data="{show: false}">
    <button x-on:click="show=!show">Toggle</button>
    <div x-show.transition.delay.1000.duration.1000="show">
        Parent
        <div x-show.transition.duration.1000="show">
            Child
        </div>
    </div>
</div>
```
- In this example root parent detects child element's transition duration and use as its transition delay in a chain so the root parent will wait 2 seconds for all nested child and will complete the transition in total of 3 seconds. 
```html
<div x-data="{show: false}">
    <button x-on:click="show=!show">Toggle</button>
    <div x-show.transition.duration.1000="show">
        Parent
        <div x-show.transition.duration.1000="show">
            Child
            <div x-show.transition.duration.1000="show">
                Sub Child
            </div>
        </div>
    </div>
</div>
```

What do you think?

@SimoTod i have mentioned this on one of your PR if you remember, i thought it would be nice to discuss.

**Comment by KevinBatdorf** [comment]
I like the `.delay` modifier approach as it gives control to the developer. 

Some complex components would break if they suddenly had the parent elements waiting though. For example, if you have a parent div that slides right out of view while the child elements fade out. Meanwhile another child is doing something else.

A third idea would be some sort of "wait on children to finish transitioning" modifier. like `.wait` maybe?

```html
<div x-show.transition.duration.1000.wait="show">
  Parent
  <div x-show.transition.duration.1000="show">
    Child
  </div>
</div>
```

**Reply by MuzafferDede** [reply]
@KevinBatdorf  I think by default parents already wait for the child's to complete the transitions. The problem is (it might be a bug) parent with transition does not wait for the child's. Right now parent and child will start the transition at same time and if child's duration is longer than parent's, parent will first fade out for example and than back to show state until child is completed, then hide immediately. (as @SimoTod mentioned down below). Which i feel it's a bit weird. And i am not sure if it will cause any breaking changes since the current version does not support this idea so not sure if anyone has such a use case.

**Comment by SimoTod** [comment]
@muzafferdede yeah, I remember it.

So, when showing an item, I don't think the parent wait (and I don't think it should wait either) for the child.
If the parent has display none, the child would be hidden so waiting 1 seconds when nothing visible happens would look like the animation is broken.

When hiding, the parent wait for the child to call the final hide step but I don't think transition are supported so I believe they currently start at the same time and the parent will just sit in a state with opacity 0 for a longer time until the child has finished, the set display to none.

Both the wait and delay seems cool but I'm not sure how difficult it would be to implement them.
delay seems more flexible to me, you could start the fade while the children is half way through or just add a voluntary delay to an existing transition).

**Reply by MuzafferDede** [reply]
@SimoTod Yeah, i know there is a step where parent has to be visible while child is doing its thing or likewise. But i feel it shouldn't be too hard to achieve a way to have a transition chain relay on a single variable since parent without transition already works. This could open ways for creating `domino effect` animations :). 

By the way, my use case is a `loading` animation where i need a to wait a child transition to finish then fade out the parent. Right now i use 2 variables to achieve this where feel the need.

**Reply by SimoTod** [reply]
the transitionDelay approach should work. Make sure you sum the delay to the timeout value. At the moment, it only uses the transition duration so the hide callback runs after 1s rather the 2s

**Reply by SimoTod** [reply]
https://github.com/alpinejs/alpine/blob/master/src/utils.js#L460 and https://github.com/alpinejs/alpine/blob/master/src/utils.js#L463

**Reply by MuzafferDede** [reply]
https://codepen.io/muzafferdede/pen/jOWdZZE
Yeah, got something working for the out animation but for the in animation child doesn't wait parent which make sense when i look at the delay times. My mind stuck at the moment :/

**Reply by SimoTod** [reply]
Yeah, if you add the delay at line 259 (JS section) the parent will wait for the children which means that you want see anything for the whole animation, that's why show, by default, doesn't wait for the children so I don't think it's really what you want.

do you need something like https://codepen.io/SimoTod/pen/yLeZKdE ? (I just forked your codepen and changed line 259)

**Reply by MuzafferDede** [reply]
~Yeah, i did few ways but couldn't get what i want. for now a stupid way coming to my mind as `delay-in.1000 delay-out.2000` so i can check in out and delay. Which i don't really like, i would love to see if alpine could do all these for me by checking parent and child durations. Not sure if i make sense :)~

> do you need something like https://codepen.io/SimoTod/pen/yLeZKdE ? (I just forked your codepen and changed line 259)

just saw this. and yeah seems this is what i need :)

**Reply by MuzafferDede** [reply]
~But i did the same thing, why mine didn't work?~
Haha, didn't notice you also changed the html... now make sense haha. I think we got ourself a PR for the transition delay :D

**Reply by SimoTod** [reply]
it's a bit weird when you interrupt a transition though

**Reply by MuzafferDede** [reply]
Yeah, i notice that too, it's waiting for ghost transitions to complete even they are not running at all. Which is expected since we are kind of hardcoding all delays and durations manually and not expecting any interruptions in chain. 

Also there is a bigger issue/pain, the PR i am working on which i think will cause all transition tests to be added transition-delay on received styles. Which means bunch of tests needs to be revised. Or we have to do some checks as if there is a delay modifier then do the 'delay' things in multiple places. Maybe i push a version to my github and we can have a look.

**Reply by MuzafferDede** [reply]
Just pushed to github 

https://github.com/alpinejs/alpine/compare/master...muzafferdede:transition-delay?expand=1

There is not much changes and we are green on existing tests. At the end will add tests for our goal once we have a solid path. 

Updated Codepen with alpine.js linked to my github alpine dist
https://codepen.io/muzafferdede/pen/jOWdZZE

**Reply by SimoTod** [reply]
I usually add tests as soon as I can since they help me understanding and developing the feature. Make sure they are not overcomplicated and they test the feature properly because Caleb also uses them when reviewing. 👍

**Reply by MuzafferDede** [reply]
Alright 👍

**Reply by MuzafferDede** [reply]
@SimoTod i am ready to PR, do you want to have a look how it is so far? 

https://github.com/alpinejs/alpine/compare/master...muzafferdede:transition-delay?expand=1

**Reply by SimoTod** [reply]
probably not too important but, if your css file is not well written and you have a delay but not a duration, you probably want the final duration variable to be 0 since no transition will be triggered anyway.

let duration = transitionDuration ? (transitionDuration + transitionDelay) * 1000 : 0;

Same for the if, if you just have a delay but not a duration, you probably want to check if there is a valid animation.

Those are edge cases, though.

**Reply by MuzafferDede** [reply]
Just want to confirm before i dig in, you know that alpine by default sets transition duration to 150ms for in and 75ms for the out animation. Even so dev's css will effect your claim? 

```js
        duration: modifierValue(modifiers, 'duration', 150),
....///
    const duration = settingBothSidesOfTransition
        ? modifierValue(modifiers, 'duration', 150)
        : modifierValue(modifiers, 'duration', 150) / 2
```

**Reply by SimoTod** [reply]
Of course I know. the transition modifier is just a shortcut to set a default transition but if you want full flexibility, you'll probably use the x-transition directives. That ones don't have a default. 
As I said, it's an edge case and it shouldn't really happen: it means that someone put a class with a delay but without a duration in a x-transition directive which is a dev mistake. Just leave it as it is, thanks, I changed my mind and it would just make the PR more complicate to review. :)

**Reply by MuzafferDede** [reply]
Oh i see your point. I believe we are fine because browsers by default does this:

```js
getComputedStyle($0).transitionDuration
"0s"

getComputedStyle($0).transitionDelay
"0s"
```
I think shouldn't be a problem. Or just like you said, it is very edge case where it's dev's mistake which alpine should not be taking care of. Better to have an issue opened and see a real use case to take care of it instead of guessing dev's mistakes :). Pushing PR now.. (#673 )

**Comment by kjjd84** [comment]
this has been an issue for 5 years now

for example a multi-step form, having each step fade out, wait, and the next one fade in is impossible with alpine without implementing a ton of bizarre hacks and workarounds because you cannot specify separate `in` and `out` delays

i also noticed that other issues for this were converted into "discussions" and PR's made to try and resolve this were just closed
